lib/timers.js

```diff
@@ -1,52 +1,110 @@
 
-// IDLE TIMEOUTS
+
+// HOW and WHY the timers implementation works the way it does.
+//
+// Since many parts of Node.js and user applications rely on timeouts, there may
+// be a very large amount of timeouts scheduled at any given time.
+// Therefore, it is very important that the timers implementation is very fast
+// and efficient.
+//
+// Note: It is suggested you first read though the lib/internal/linkedlist.js
+// linked list implementation. How it works is somewhat counter-intuitive to
+// most JavaScript code, and timers depend on it extensively.
 //
-// Because often many sockets will have the same idle timeout we will not
-// use one timeout watcher per item. It is too much overhead.  Instead
-// we'll use a single watcher for all sockets with the same timeout value
-// and a linked list. This technique is described in the libev manual:
-// http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#Be_smart_about_timeouts
-
-// Object containing all lists, timers
-// key = time in milliseconds
-// value = list
-var lists = {};
-
-
-// call this whenever the item is active (not idle)
-// it will reset its timeout.
-// the main function - creates lists on demand and the watchers associated
-// with them.
+// In order to be as efficient as possible, the implementation does as little as
+// possible, and only when required.
+//
+// Object maps are kept which contain linked lists keyed by their duration in
+// milliseconds.
+// The linked lists within also have some meta-properties, one of which is a
+// TimerWrap C++ handle, which makes the call after the duration to process the
+// list it is attached to.
+//
+// With this, virtually constant-time insertion (append), removal, and timeout
+// is possible in the JavaScript layer. Any one list of timers is able to be
+// sorted by just appending to it because all timers within share the same
+// duration. Therefore, any timer added later will always have been scheduled to
+// timeout later, thus only needing to be appended. Any less-than constant
+// overhead is contained within the TimerWrap's inherently faster libuv binding.
+// There is, however, a minor overhead from accessing the specific list from the
+// map, or creating a new one, but it is trivial in comparison.
+// Removal from an object-property linked list is also constant-time as can be
+// seen in the lib/internal/linkedlist.js implementation.
+// Timeouts only need to process any timers due to currently timeout, which will
+// always be at the beginning of the list for reasons stated above. Any timers
+// after the first one encountered that does not yet need to timeout will also
+// always be due to timeout at a later time.
+
+
+// Object maps containing linked lists of timers, keyed and sorted by their
+// duration in milliseconds.
+// - key = time in milliseconds
+// - value = linked list
+const refedLists = {};
+const unrefedLists = {};
+// The difference between these two objects is that the former contains timers
```

Comment:
Nitpicking: in general comments are written above the code that they describe. Writing them after the code they mention can be confusing as it could suggest that they used to refer to code that has been deleted.